Change to logfmt format

Thomasdezeeuw opened the pull request:
The logfmt doesn't really have a standard, but it is described here:
https://www.brandur.org/logfmt, along with its best practices.

The new implementation uses vectored writing, reducing the amount of
copying when formatting the log record.

TODO:
* [ ] Use vectored I/O in key-values of the record.
* [ ] Benchmark copying into a single buffer vs. using more vectored I/O. I.e. at what point is copying *x* bytes cheaper than using another `iovec`.

Thomasdezeeuw commented:
Blocked by https://github.com/sval-rs/value-bag/issues/15.

KodrAus commented on src/format.rs:
```diff
@@ -1,97 +1,247 @@
+        self.0.push(b' ');
+        self.0.extend_from_slice(key.as_str().as_bytes());
+        self.0.push(b'=');
+        if let Some(value) = value.to_borrowed_str() {
```
These `to_*` methods aren't really meant for exhaustively visiting the value. So instead of writing something like:

```rust
if let Some(v) = value.to_u64() {
  ..
} else if let Some(v) = value.to_u32() {
  ..
} else if ..
```

you can either use the `kv_unstable_sval` feature along with [`sval` in no-std](https://docs.rs/sval/1.0.0-alpha.4/sval/) or the `kv_unstable_serde` feature along with [`serde` in std](https://docs.rs/serde) to visit the value. Our documentation is still very sparse so this is really not obvious! If you'd like I can put together an example of what this would look like.

It looks like this should be about the same as `write!(self.1, "{}={:?}", key, value)` though, was there a reason you needed to replace the previous code calling `write!(self.1, "{}={}", key, value)`? Was there something different about the format that meant you had to throw the whole `Debug` format out?

Thomasdezeeuw replied:
> These `to_*` methods aren't really meant for exhaustively visiting the value. So instead of writing something like:

Ah, I didn't get that from the documentation.

> you can either use the `kv_unstable_sval` feature along with [`sval` in no-std](https://docs.rs/sval/1.0.0-alpha.4/sval/) or the `kv_unstable_serde` feature along with [`serde` in std](https://docs.rs/serde) to visit the value. Our documentation is still very sparse so this is really not obvious! If you'd like I can put together an example of what this would look like.

I haven't looked too much as sval, but doesn't that require the structure to implement `sval::value::Value`? Same question for serde. Basically I'm looking the get the primitive type from the `Value`, would it perhaps be possible for valuebag/log to expose `valuabag::internal::Primitive` in some way?

> It looks like this should be about the same as `write!(self.1, "{}={:?}", key, value)` though, was there a reason you needed to replace the previous code calling `write!(self.1, "{}={}", key, value)`? Was there something different about the format that meant you had to throw the whole `Debug` format out?

Mainly I'm looking to optimise logging by using more vectored I/O and reducing the overhead of formatting, e.g. I'm also using the `itoa` crate rather than std library's formatting.

KodrAus replied:
The `Value` type will already implement `sval::Value` or `serde::Serialize` so all we need to do is write a `Stream` or `Serializer` that will format the way you want.

When I get a chance over the weekend I’ll put something together to show you.

KodrAus replied:
[Here's an example of using `sval` to visit the value](https://github.com/Thomasdezeeuw/std-logger/compare/logfmt...KodrAus:logfmt).

It's come up a few times that users who just want to visit a few primitive datatypes end up chaining `to_*` methods though, so I've also opened https://github.com/rust-lang/log/issues/440 to provide a simple zero-dependency visitor API for `log::kv::Value`s that should cover this case without needing to bring in `sval` directly.

Thomasdezeeuw replied:
@KodrAus thanks for the example. One more question, If I have a `kv::key::Source` created using `serde` can I still use `sval::Stream` for this source? I.e. would the correct `Stream::u64`, etc., functions be called? Or would it use the display implementation?

KodrAus replied:
@Thomasdezeeuw It would, yes. One of the goals of the design is that you get the closest possible representation of your value without having to worry about what was passed in in the first place.

So if a value is created using `serde` that serializes as a `u64` or struct or sequence then you'll get an appropriate `u64` or map or sequence from `sval`. You'll also get something that looks like typical `#[derive(Debug)]` if you decide to format it using `{:?}` or `{}`.

KodrAus commented on src/format.rs:
```diff
@@ -1,97 +1,247 @@
+        self.0.push(b'=');
+        if let Some(value) = value.to_borrowed_str() {
+            self.0.push(b'\"');
+            self.0.extend_from_slice(value.as_bytes());
```
Just wondering, is escaping a concern here if the value contains nested `"` characters?

Thomasdezeeuw replied:
Yes it should be, for now I'm just sort of ignoring it. I don't really want to check the string, so I think I want to solve in the consumer of the logs. But I haven't made up my mind yet.

Thomasdezeeuw committed:
Fix key_value example

It logged the url twice, rather then the method.

Thomasdezeeuw committed:
Change to logfmt format

The logfmt doesn't really have a standard, but it is described here:
https://www.brandur.org/logfmt, along with its best practices.

The new implementation uses vectored writing, reducing the amount of
copying when formatting the log record.

Thomasdezeeuw committed:
Log correct file and line in panic hook

Since we are logging a panic that happened elsewhere we can use the file
and line from the panic information. This also removes the module as
that can't be obtained.

Thomasdezeeuw committed:
Update the format documentation

And some additional small things.

Thomasdezeeuw committed:
Add documentation for the nightly feature

Thomasdezeeuw commented:
Some follow up issues: #34, #35.
